### PR TITLE
Add fractional seconds support components bag

### LINKED ISSUE #1360: Add Fractional Seconds to the components bag
Depends on #485.

Fractional seconds should be added to the DateTimeFormat component in order to be compatible with ECMA-402's components bag.

### PR DESCRIPTION
**Pull request by dminor**
Fixes #1360

### CODE REVIEW

File: components/datetime/src/skeleton/helpers.rs
Comment (gregtatum) [inline components/datetime/src/skeleton/helpers.rs:435]: Question: Could this retain the fractional second index and save a second loop through requested fields? See my `Thought:` above.

I'm listing this as a question, since it could be used in an optional perf follow-up, as listed above.

File: components/datetime/src/skeleton/helpers.rs
Comment (gregtatum) [inline components/datetime/src/skeleton/helpers.rs:330]: thought (follow-up?): For what it's worth, this is iterating through the requested fields another time, and incurs the cost of another loop through the field. This occurs for every pattern lookup. I believe an index to the seconds field could be returned from the skeleton matching algorithm to skip this loop.

The common case here is for there not to be a fractional second, so this would remove a loop from the common case, and potentially save time. I'll leave it up to you if you think this is worth filing as a follow-up.

File: components/datetime/src/skeleton/helpers.rs
Comment (gregtatum) [inline components/datetime/src/skeleton/helpers.rs:334]: Thought: This is incurring an additional allocation. Nothing comes to mind to optimize it out. It does seem odd to have to go through the `vec` and `ZeroVec` allocations to do any mutation.

First the ZeroVec is turned into a vec: `self.iter().collect()`
Next it's converted back to a ZeroVec: `ZeroVec::alloc_from_slice`

File: components/datetime/src/skeleton/helpers.rs
Comment (gregtatum) [inline components/datetime/src/skeleton/helpers.rs:329]: Thought (this is an existing issue, so you are in no obligation to do any action on it, I just wanted to bring it up): 

Having `icu_datetime::pattern::reference::pattern::Pattern` and `icu_datetime::pattern::runtime::pattern::Pattern` makes this super confusing to review, as opposed to having a struct named `ReferencePattern` and `RuntimePattern`. `Pattern` is pretty much impossible to grep in the codebase. There's even another `Pattern` struct at `utils/pattern/src/pattern/mod.rs`.

In addition, since these aren't publicly exposed, there's not documentation on the structs themselves, which leads to the awkward situation in the editor where no documentation is provided:

image

This is mostly a bad code reading experience, which makes it hard to reason with some code changes like the ones in this function that are moving between reference and runtime patterns.

File: components/datetime/tests/fixtures/tests/components.json
Comment (gregtatum) [inline components/datetime/tests/fixtures/tests/components.json:38]: Suggestion: As stated above, a test like this one with `"fractional_second": 0` would be nice to have, since it's a bit weird and expressible in the type system.

Comment (gregtatum) [inline components/datetime/tests/fixtures/tests/components.json:38]: Suggestion: A test with `fractional_second": 255` would be nice as well.

**Review by gregtatum — APPROVED** [review]
Thanks, the patch looks good to me for approval.

I left mostly questions and thoughts, the only suggestions that should be handled before merging is the test coverage for `0` and `255` fractional seconds.

File: components/datetime/src/skeleton/helpers.rs
Comment (dminor) [reply components/datetime/src/skeleton/helpers.rs:330]: I filed https://github.com/unicode-org/icu4x/issues/1888 for the performance optimizations you suggested. Thanks for the ideas :)

File: components/datetime/src/skeleton/helpers.rs
Comment (dminor) [reply components/datetime/src/skeleton/helpers.rs:334]: Yes, I think we're stuck with this... `ZeroVec` works best when backed by static data.

File: components/datetime/src/skeleton/helpers.rs
Comment (dminor) [reply components/datetime/src/skeleton/helpers.rs:329]: Filed https://github.com/unicode-org/icu4x/issues/1889 for this.

**Comment by dminor** [comment]
The codecov failure might be due to a bad token:
```
->  Pinging Codecov
https://codecov.io/upload/v4?package=github-action-v1.5.2-1.0.3&token=secret&branch=issue-1360-fractional-component&commit=3ae3c2e68c2653a524c5103e93939a3a4d935393&build=2333577834&build_url=http%3A%2F%2Fgithub.com%2Funicode-org%2Ficu4x%2Factions%2Fruns%2F2333577834&name=&tag=&slug=unicode-org%2Ficu4x&service=github-actions&flags=&pr=1873&job=Coverage&cmd_args=n,F,Q,Z,f,C
{'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
```

File: components/datetime/tests/fixtures/tests/components.json
Comment (gregtatum) [inline components/datetime/tests/fixtures/tests/components.json:32]: Question: Is it expected that a fractional second of 0 should default to 3 instead? This seems a little odd behavior.

**Review by gregtatum — APPROVED** [review]

**Pull request merged**
